readthedocs/projects/tasks/builds.py

```diff
@@ -135,14 +139,31 @@ def on_failure(self, exc, task_id, args, kwargs, einfo):
                 'There was an error with the repository.',
                 exc_info=True,
             )
+        elif isinstance(exc, SyncRepositoryLocked):
+            log.warning(
+                "Skipping syncing repository because there is another task running."
+            )
         else:
             # Catch unhandled errors when syncing
             log.exception('An unhandled exception was raised during VCS syncing.')
 
     def after_return(self, status, retval, task_id, args, kwargs, einfo):
         clean_build(self.data.version)
 
+    def _rate_limit(self):
+        """
+        Rate limit the task per-project.
+
+        Allow to execute 1 task per minute.
+        """
+        lock_id = f"{self.name}-lock-{self.data.project.slug}"
+        locked = not cache.add(lock_id, oid, 60)
+        if locked:
+            raise SyncRepositoryLocked
```

Comment:
we should release the lock once the task ends, otherwise we will lose data.

- Create branch
- Sync task is done
- Create another branch after 30 seconds
- Task is locked, the new branch isn't added

Reply:
releasing the lock can be done at a celery handler when the task finishes, don't think that should be too complex to add.

Reply:
If we release the lock we lost the feature this PR implements.

Reply:
Another approach could be to "do not execute the task if there is another one with the same `project.id` queued". So, if 1k tasks are triggered for the same project, we will skip 999 of them and execute the last one. This SO answer could help on that https://stackoverflow.com/a/46453503

However, it involves using Celery internals to check the queued tasks, which is not _super fast_.

Reply:
I think the edge case mentioned here is valid and we are not covering it. However, I doubt we will hit it frequently, because _besides_ the steps you described, _that particular branch_ created has to be a version the user wants to activate as a documentation version.

So, to me, it's very unlikely this will happen often. Besides, the solution/workaround is as simple as triggering a build.

Reply:
That said, I'm not sure it is worth the complexity added and I think we can start testing this simple approach and revert it if we find it does not work as we expect.

Reply:
Wouldn't each of the 999 get skipped by this logic, since they would presumably be read off the queue while another worker while celery is actively working on it?

Reply:
> Wouldn't each of the 999 get skipped by this logic, since they would presumably be read off the queue while another worker while celery is actively working on it?

If by _this logic_ you refer to Santos' idea (release the lock after finishing the task): No. More than 1 of them will be executed. Now sure how many, it will depend on how many instances we have at the time the task was triggered. If 1k tasks are triggered with 10 workers, 1 will take and execute the task. The other 9 will take and discard it while the first task is executed. However, after the first task finishes, another one will be exected and 9 more will be discarded while the new one is executed. The ratio of tasks executed will be ~`total tasks triggered / worker instances`

Anyways, there is not too much difference between either approach when we have many workers running in parallel. Santos' idea behaves worse when we have just a few workers, tho. So, I'm happy to change it as you prefer. Let me know.

Reply:
@humitos I don't understand why a worker wouldn't read and discard multiple of the tasks? It seems that a worker would just keep reading the tasks from the queue and discarding them, presumably doing multiple of these per second?

It feels like we're all just debating without actually knowing what will happen, but I'm guessing it will lead to issues. I'm imagining a situation where I:

* Push a change to a new branch, which triggers sync_repos
* I realize I wanted a different branch name, so I do `git co -b new-branch && git push`
* If that happens within 60 seconds of the original push, won't `new-branch` not get synced?

That seems like a pretty common edge case, or at least one that users will actually hit?

I'd much prefer some approach that implements the logic:

* Execute at most 1 sync every minute per project
* Never discard the last sync operation for a project, so we always "catch up" on the sync

That seems to be what the SO question you were pointing to is implementing, but I agree it's pretty complicated, but I think that's the correct solution to this problem and hopefully doesn't have edge cases, if we can implement it effectively.

Reply:
@ericholscher 

> I don't understand why a worker wouldn't read and discard multiple of the tasks? It seems that a worker would just keep reading the tasks from the queue and discarding them, presumably doing multiple of these per second?

Yes. This is correct.

> I'd much prefer some approach that implements the logic:
>    Execute at most 1 sync every minute per project
>    Never discard the last sync operation for a project, so we always "catch up" on the sync

This solution is the ideal one, but I don't know how to implement it in a reliable and simple way.

> That seems to be what the SO question you were pointing to is implementing, but I agree it's pretty complicated

The SO that I linked shows how to inspect tasks that are running. We should modify it to make usage of `project.slug`, tho. I'm not sure the edge cases we are trying to avoid is worth this effort and probably may bring other issues due to its complexity.

I think that Santos' ideas or mine are good enough to move forward and any of them are better than what we currently have.

Reply:
I'd prefer @stsewd approach I think. Using extra resources is better than missing syncing, so reducing the amount of time that we're missing syncing from is a good thing.